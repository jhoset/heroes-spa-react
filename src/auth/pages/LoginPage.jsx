import { useContext } from "react";
import { useNavigate } from "react-router-dom"
import { AuthContext } from "../context/AuthContext";

export const LoginPage = () => {

    const navigate = useNavigate();
    const { loginFn } = useContext(AuthContext);

    const onLogin = () => {
        const lastPath = localStorage.getItem('lastPath') || '/'
        loginFn("Sub Zero");
        navigate(lastPath, {
            replace: true
        })
    }

    return (
        <div className="container mt-5">
            <h1> Login </h1>
            <hr />
            <button
                onClick={onLogin}
                type="submit"
                className="btn btn-primary">Login</button>
        </div>
    )

}