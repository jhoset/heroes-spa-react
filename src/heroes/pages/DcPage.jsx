import { HeroList } from "../components/HeroList"

export const DcPage = () => {
    return (
        <>
            <div className="d-flex gap-2 align-items-center">
                <img height="40px" src={`/assets/heroes/dc-logo.png`} alt="" />
                <h1> Heroes </h1>
            </div>
            <hr />

            <HeroList publisher={"DC Comics"} />
        </>
    )
}