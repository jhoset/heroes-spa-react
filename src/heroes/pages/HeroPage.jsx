import { Navigate, useNavigate, useParams } from "react-router-dom"
import { getHeroById } from "../helpers";
import { useMemo } from "react";

export const HeroPage = () => {
    const navigate = useNavigate();
    const { id } = useParams();
    const hero =  useMemo(() => getHeroById(id), [id]) ;

    const onNavigateBack = () => {
        navigate(-1)
    }


    if (!hero) {
        return <Navigate to={"/marvel"} />
    }


    return (
        <div className="row pt-4 rounded-5 animate__animated animate__fadeIn">
            <div className="col-4 animate__animated animate__fadeInLeft">
                <img src={`/assets/heroes/${id}.jpg`} alt={hero.superhero} className="img-thumbnail" />
            </div>

            <div className="col-8 d-flex flex-column justify-content-between animate__animated animate__fadeIn">
                <h2> {hero.superhero}</h2>
                <ul className="list-group list-group-flush gap-2 ">
                    <li className="list-group-item rounded-2 "> <b>Alter ego:</b> {hero.alter_ego} </li>
                    <li className="list-group-item rounded-2 "> <b>Publisher:</b> {hero.publisher} </li>
                    <li className="list-group-item rounded-2 "> <b>First Appearance:</b> {hero.first_appearance} </li>
                </ul>
                <div>
                <h5 className="mt-3">
                    Characters
                </h5>
                <p>
                    {hero.characters}
                </p>
                </div>

                <button onClick={onNavigateBack} className="btn btn-outline-info w-100">
                    Back
                </button>
            </div>
        </div>
    )

}