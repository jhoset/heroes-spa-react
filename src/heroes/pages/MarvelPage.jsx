import { HeroList } from "../components/HeroList"

export const MarvelPage = () => {
    return (
        <>
            <div className="d-flex gap-2 align-items-center">
                <img height="40px" src={`/assets/heroes/marvel-logo.png`} alt="" />
                <h1> Heroes </h1>
            </div>

            <hr />

            <HeroList publisher={"Marvel Comics"} />
        </>
    )
}