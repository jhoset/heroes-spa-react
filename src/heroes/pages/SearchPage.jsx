import { useLocation, useNavigate } from "react-router-dom"
import queryString from 'query-string';
import { useForm } from "../../hooks/useForm"
import { HeroCard } from "../components/HeroCard"
import { getHeroesByName } from "../helpers";
export const SearchPage = () => {
    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);

    const heroes = getHeroesByName(q);

    const { searchText, onInputChange } = useForm({
        searchText: q
    })

    const onSearch = (event) => {
        event.preventDefault();
        // if (!searchText.trim().length) return;
        navigate(`?q=${searchText.trim()}`);
    }



    return (
        <>
            <h1> Search </h1>
            <hr />
            <div className="row ">
                <div className="col-md-5">
                    <label htmlFor="">Hero Name:</label>
                    <form onSubmit={onSearch} className="mt-1">
                        <input onChange={onInputChange} value={searchText} type="text" autoComplete="off" name="searchText" placeholder="Search Hero" className="form-control" />
                        <button type="submit" className="btn btn-outline-warning mt-3 w-100">Search</button>
                    </form>
                </div>
            </div>
            <hr />
            <h2> Results:</h2>
            {
                q == '' && (
                <div className="alert alert-warning animate__animated animate__fadeIn">
                    Please enter a name to search a hero <b>{q}</b>
                </div>)
            }
                        {
                q !== '' && !heroes.length && (
                <div className="alert alert-danger animate__animated animate__fadeIn">
                    No records found with name: <b>{q}</b>
                </div>)
            }
            <div className="row row-cols-1 row-cols-md-2 row-cols-lg-3 row-cols-xl-4 g-4">
                {
                    heroes.map(hero => (
                        <HeroCard key={hero.id} hero={hero} ></HeroCard>
                    ))
                }
            </div>

        </>
    )

}