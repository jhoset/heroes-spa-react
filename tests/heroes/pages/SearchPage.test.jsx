import { fireEvent, render, screen } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"
import { SearchPage } from "../../../src/heroes/pages/SearchPage"

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate
}))

describe("Tests on <SearchPage />", () => {


    beforeEach(() => {
        jest.clearAllMocks();
    })

    test("Should display correctly with default values ", () => {
        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot();
    })

    test("Should display Batman and input with queryString", () => {
        const queryParameter = "batman"
        render(
            <MemoryRouter initialEntries={[`/search?q=${queryParameter}`]} >
                <SearchPage />
            </MemoryRouter>
        )

        // screen.debug();
        const input = screen.getByRole("textbox");
        expect(input.value).toBe(queryParameter);
        const heroName = screen.getByLabelText("heroName");
        expect(heroName.innerHTML.toString().toLowerCase().trim()).toContain(queryParameter);
        const warningAlert = screen.queryByLabelText("alertWarning");
        const dangerAlert = screen.queryByLabelText("alertDanger");
        expect(warningAlert).toBe(null);
        expect(dangerAlert).toBe(null);

    })

    test("Should display an error if hero not found", () => {
        const queryParameter = "Zero";
        render(
            <MemoryRouter initialEntries={[`/search?q=${queryParameter}`]}>
                <SearchPage />
            </MemoryRouter>
        )

        const dangerAlert = screen.queryByLabelText("alertDanger");
        expect(dangerAlert).toBeTruthy();
        
    })

    test("Should call navigate to new screen", () => {
        const queryParameter = "Zero";
        render(
            <MemoryRouter initialEntries={[`/search?q=${queryParameter}`]}>
                <SearchPage />
            </MemoryRouter>
        )
        const searchButton = screen.queryByRole("button");
        fireEvent.click(searchButton);
        expect(mockUseNavigate).toHaveBeenCalledWith(`?q=${queryParameter}`);
    })


})