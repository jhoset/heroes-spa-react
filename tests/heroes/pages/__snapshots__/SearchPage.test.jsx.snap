// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <SearchPage /> Should display correctly with default values  1`] = `
<div>
  <h1>
     Search 
  </h1>
  <hr />
  <div
    class="row "
  >
    <div
      class="col-md-5"
    >
      <label
        for=""
      >
        Hero Name:
      </label>
      <form
        class="mt-1"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search Hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-warning mt-3 w-100"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
  </div>
  <hr />
  <h2>
     Results:
  </h2>
  <div
    aria-label="alertWarning"
    class="alert alert-warning animate__animated animate__fadeIn"
  >
    Please enter a name to search a hero 
    <b />
  </div>
  <div
    class="row row-cols-1 row-cols-md-2 row-cols-lg-3 row-cols-xl-4 g-4"
  />
</div>
`;
